mlir/lib/Dialect/Rock/Transforms/Regularize.cpp

```diff
@@ -147,187 +146,452 @@ struct RegularizeGenericRewritePattern
+               dyn_cast_or_null<ViewLikeOpInterface>(inp.getDefiningOp()))
+      inp = viewOp.getViewSource();
+
+    if (isa<BlockArgument>(inp)) {
+      auto arg = dyn_cast<BlockArgument>(inp);
+      auto shape = cast<ShapedType>(inp.getType());
+      int64_t argSize = shape.getNumElements();
+      if (inputIdx == 0 || argSize > majorTensorSize ||
+          (argSize == majorTensorSize && argIdx > arg.getArgNumber())) {
+        majorTensorIdx = inputIdx;
+        majorTensorSize = argSize;
+        argIdx = arg.getArgNumber();
       }
-      inputIdx++;
     }
-    if (majorTensorIdx >= 0)
-      lgop->setAttr("rock.majorTensorNumber",
-                    IntegerAttr::get(IndexType::get(ctx), majorTensorIdx));
   }
-  return;
+  if (majorTensorIdx >= 0)
+    lgop->setAttr("rock.majorTensorNumber",
+                  IntegerAttr::get(IndexType::get(ctx), majorTensorIdx));
 }
 
 ////////////////////////////////////////////////////////////////////////
-////  Push Transforms Over alloc to writer
+////  Classify buffer writes into input and output fusion chains
 ////////////////////////////////////////////////////////////////////////
-struct PushTransformsUpRewritePattern
-    : public OpRewritePattern<memref::AllocOp> {
-  using OpRewritePattern<memref::AllocOp>::OpRewritePattern;
-
-  /////////////////////////////////////////////////////////////////////
-  static bool isFusorOp(Operation *useOp, Value viewedBuffer) {
-    if (auto reduceOp = dyn_cast<rock::ReduceOp>(useOp))
-      return reduceOp.getOut() == viewedBuffer;
-    return isa<rock::GridwiseGemmOp, rock::GridwiseGemmAccelOp,
-               rock::GridwiseAttentionAccelOp, rock::ThreadwiseWriteAllOp>(
-        useOp);
-  }
 
-  static bool collectChain(Value result, Operation *forwOp,
-                           SmallVector<Operation *> &chain) {
-    while (auto top = dyn_cast<rock::TransformOp>(forwOp)) {
-      result = top.getResult();
-      if (!result.hasOneUse()) {
-        // currently restricted to 1 reader
-        LLVM_DEBUG(llvm::dbgs() << "multiple readers on transform\n");
-        return false; // TODO: fix when encountered
-      }
-      chain.push_back(forwOp);
-      forwOp = (*result.getUses().begin()).getOwner();
+namespace {
+struct TransformPushState {
+  llvm::SmallVector<memref::AllocOp> worklist;
+  llvm::SmallPtrSet<OpOperand *, 8> inputFusionWrites;
+  llvm::SmallPtrSet<OpOperand *, 8> outputFusionWrites;
+};
+} // end namespace
+
+/// Given an operand that reads a buffer, collect the writer of thet buffer and,
+/// recursively, all the writers of the buffers that writer reads, and so on.
+/// This is meant to collect all the buffer writes involved in an input fusion.
+static void collectInputFusionWrites(OpOperand *reader,
+                                     const BufferDependencyAnalysis &bufferDeps,
+                                     TransformPushState &state) {
+  std::optional<memref::AllocOp> maybeBuffer = bufferDeps.getReadBuffer(reader);
```

Comment:
so yea, whose operand that would not be a buffer in this case ?

Reply:
Function argument isn't a buffer for these purposes

Might need a s/buffer/alloc/ but also `BufferDependencyAnalysis` is focused on allocs so